libraries/AP_RCProtocol/AP_RCProtocol_FPort2.cpp

```diff
@@ -0,0 +1,264 @@
+      send it in the next call, this prevents corruption of
+      status text messages
+     */
+    if (!telem_data.available) {
+        if (!AP_Frsky_Telem::get_telem_data(telem_data.frame, telem_data.appid, telem_data.data)) {
+            return;
+        }
+        telem_data.available = true;
+    }
+    /*
+      check that we haven't been too slow in responding to the new
+      UART data. If we respond too late then we will corrupt the next
+      incoming control frame
+     */
+    uint64_t tend = uart->receive_time_constraint_us(1);
+    uint64_t now = AP_HAL::micros64();
+    uint64_t tdelay = now - tend;
+    if (tdelay > 2500) {
+        // we've been too slow in responding
+        return;
+    }
+
+    uint8_t buf[10];
+
+    buf[0] = 0x08;
+    buf[1] = frame.type;
+    buf[2] = telem_data.frame;
+    buf[3] = telem_data.appid & 0xFF;
+    buf[4] = telem_data.appid >> 8;
+    memcpy(&buf[5], &telem_data.data, 4);
+    buf[9] = crc_sum8(&buf[1], 8);
+    
+    uart->write(buf, sizeof(buf));
+
+    // get fresh telem_data in the next call
+    telem_data.available = false;
+#endif
+}
+
+/*
+  process a FPort2 input pulse of the given width
+ */
+void AP_RCProtocol_FPort2::process_pulse(uint32_t width_s0, uint32_t width_s1)
+{
+    if (have_UART()) {
+        // if we can use a UART we would much prefer to, as it allows
+        // us to send SPORT data out
+        return;
+    }
+    uint32_t w0 = width_s0;
+    uint32_t w1 = width_s1;
+    if (inverted) {
+        w0 = saved_width;
+        w1 = width_s0;
+        saved_width = width_s1;
+    }
+    uint8_t b;
+    if (ss.process_pulse(w0, w1, b)) {
+        _process_byte(ss.get_byte_timestamp_us(), b);
+    }
+}
+
+// support byte input
+void AP_RCProtocol_FPort2::_process_byte(uint32_t timestamp_us, uint8_t b)
+{
+    const bool have_frame_gap = (timestamp_us - byte_input.last_byte_us >= 2000U);
+
+    byte_input.last_byte_us = timestamp_us;
+
+    if (have_frame_gap) {
+        // if we have a frame gap then this must be the start of a new
+        // frame
+        byte_input.ofs = 0;
+    }
+
+    if (byte_input.ofs == 0) {
+        switch (b) {
+        case FRAME_LEN_8:
+            byte_input.control_len = 16;
+            chan_count = 8;
+            byte_input.is_downlink = false;
+            break;
+        case FRAME_LEN_16:
+            byte_input.control_len = 27;
+            chan_count = 16;
+            byte_input.is_downlink = false;
+            break;
+        case FRAME_LEN_24:
+            byte_input.control_len = 37;
+            chan_count = 24;
+            byte_input.is_downlink = false;
+            break;
+        case FRAME_LEN_DOWNLINK:
+            byte_input.control_len = 10;
+            byte_input.is_downlink = true;
+            break;
+        default:
+            // definately not FPort2, missing header byte
+            return;
+        }
+    }
+
+    if (byte_input.ofs == 1) {
```

Comment:
```suggestion
    } else if (byte_input.ofs == 1) {
```

Reply:
i personally find it easier to read without the else